src/pm-conversations/pmConversationsSelectors.js

```diff
@@ -51,10 +67,114 @@ export const getRecentConversations: Selector<PmConversationData[]> = createSele
          end up converted to strings, so this access with string keys works.  We should probably use
          a Map for this and similar maps. */
         unreadPms[recipient.ids] || unreadHuddles[recipient.ids],
+    })),
+);
+
+/**
+ * Legacy implementation of {@link getRecentConversations}. Computes an
+ * approximation to the set of recent conversations, based on the messages we
+ * already know about.
+ */
+const getRecentConversationsLegacyImpl: Selector<PmConversationData[]> = createSelector(
+  getOwnUser,
+  getPrivateMessages,
+  getAttachUnread,
+  (ownUser: User, messages: Message[], attachUnread: UnreadAttacher): PmConversationData[] => {
+    const items = messages.map(msg => ({
+      ids: pmUnreadsKeyFromMessage(msg, ownUser.user_id),
+      recipients: normalizeRecipientsSansMe(msg.display_recipient, ownUser.email),
+      msgId: msg.id,
     }));
+
+    const sortedByMostRecent = collateByRecipient(items);
+
+    return attachUnread(sortedByMostRecent);
   },
 );
 
+/**
+ * Modern implementation of {@link getRecentConversations}. Returns exactly the
+ * most recent conversations. Requires server-side support.
+ */
+const getRecentConversationsImpl: Selector<PmConversationData[]> = createSelector(
+  getOwnUser,
+  getAllUsersById,
+  getRecentPrivateConversations,
+  getAttachUnread,
+  (
+    ownUser: User,
+    allUsers: Map<number, UserOrBot>,
+    recentPCs: RecentPrivateConversation[],
+    attachUnread: UnreadAttacher,
+  ) => {
+    const getEmail = (id: number): string => {
+      const user = allUsers.get(id);
+      if (user) {
+        return user.email;
+      }
+      throw new Error('getRecentConversations: unknown user id');
+    };
+
+    const recipients = recentPCs.map(conversation => {
+      const userIds = conversation.user_ids.slice();
+      if (userIds.length !== 1) {
+        userIds.push(ownUser.user_id);
+        userIds.sort((a, b) => a - b);
+      }
+
+      return {
+        ids: userIds.join(','),
+        recipients: normalizeRecipientsSansMe(
+          userIds.map(id => ({ email: getEmail(id) })),
+          ownUser.email,
+        ),
```

Comment:
Oof. I think this ends up having the right behavior... but it sure is confusing from a #4035 perspective. Note that we're giving this function a different version of "the users for this PM conversation" than we do at its other call site. And that not all possible versions of that would work -- notably if we used `conversation.user_ids` directly, that would not work.

At a minimum I'd want a comment here explaining why this ends up with the right set of users (and what the right set of users is supposed to be.)

Better would be another small function in `recipients.js` with its own name and jsdoc.

A good followup would be to eliminate this `recipients` property entirely. It flows to just a handful of places, all under our direct control.